File: frontend-react/src/utils/passwordPolicy.js
/**
 * Hexaq
 * 계층: Utils
 * 객체: passwordPolicy
 * 책임: 도메인 규칙. HTTP·DB 모름
 * 문서: [docs/features/overview.md](../../../docs/features/overview.md) · [docs/technical/01-architecture.md](../../../docs/technical/01-architecture.md)
 */
export const PASSWORD_RULE_TEXT = '8~20자, 영문+숫자+특수문자 조합'

export function isStrongPassword(password) {
  if (!password || password.length < 8 || password.length > 20) {
    return false
  }
  if (/\s/.test(password)) {
    return false
  }
  return /[A-Za-z]/.test(password) && /[0-9]/.test(password) && /[^A-Za-z0-9]/.test(password)
}

export function passwordResetHint(password) {
  if (!password) {
    return ''
  }
  const missing = []
  if (password.length < 8 || password.length > 20) {
    missing.push('8~20자')
  }
  if (!/[A-Za-z]/.test(password)) {
    missing.push('영문')
  }
  if (!/[0-9]/.test(password)) {
    missing.push('숫자')
  }
  if (/\s/.test(password) || !/[^A-Za-z0-9]/.test(password)) {
    missing.push('특수문자')
  }
  if (missing.length === 0) {
    return ''
  }
  return `${PASSWORD_RULE_TEXT}이 아닙니다(${missing.join(', ')} 필요). 조건을 맞춰 다시 입력해 주세요.`
}

File: src/main/java/com/edu/springboot/domain/member/PasswordPolicy.java
package com.edu.springboot.domain.member;

import java.util.regex.Pattern;

/**
 * Hexaq
 * 계층: Domain
 * 객체: PasswordPolicy
 * 책임: 도메인 규칙. HTTP·DB 모름
 * 문서: [docs/features/01-signup-verification.md](../../../../../../../../docs/features/01-signup-verification.md) · [docs/security/06-identity-verification.md](../../../../../../../../docs/security/06-identity-verification.md) · [docs/technical/01-architecture.md](../../../../../../../../docs/technical/01-architecture.md)
 */
public class PasswordPolicy {

	private static final int MIN_LENGTH = 8;
	private static final int MAX_LENGTH = 20;
	private static final Pattern LETTER = Pattern.compile("[A-Za-z]");
	private static final Pattern DIGIT = Pattern.compile("[0-9]");
	private static final Pattern SPECIAL = Pattern.compile("[^A-Za-z0-9]");
	private static final String RULE_MESSAGE = "비밀번호는 8~20자이며 영문, 숫자, 특수문자를 모두 포함해야 합니다.";

	public boolean matches(String password) {
		if (password == null || password.length() < MIN_LENGTH || password.length() > MAX_LENGTH) {
			return false;
		}
		if (password.chars().anyMatch(Character::isWhitespace)) {
			return false;
		}
		return LETTER.matcher(password).find()
			&& DIGIT.matcher(password).find()
			&& SPECIAL.matcher(password).find();
	}

	public String ruleMessage() {
		return RULE_MESSAGE;
	}
}
